virtualizarr/manifests/manifest.py

```diff
@@ -56,25 +62,131 @@ class ChunkManifest(BaseModel):
             "0.1.1": {"path": "s3://bucket/foo.nc", "offset": 400, "length": 100},
         }
 
+    using the .__init__() and .dict() methods, so users of this class can think of the manifest as if it were a dict mapping zarr chunk keys to byte ranges.
 
-    See the chunk manifest SPEC proposal in https://github.com/zarr-developers/zarr-specs/issues/287 .
+    (See the chunk manifest SPEC proposal in https://github.com/zarr-developers/zarr-specs/issues/287.)
 
     Validation is done when this object is instatiated, and this class is immutable,
-    so it's not possible to have a ChunkManifest object that does not represent a complete valid grid of chunks.
+    so it's not possible to have a ChunkManifest object that does not represent a valid grid of chunks.
     """
 
-    model_config = ConfigDict(frozen=True)
+    _paths: np.ndarray[Any, np.dtype[np.dtypes.StringDType]]
```

Comment:
So originally I was going to wait for numpy2.0 before merging this, but actually I could just merge it before that by using a fixed-length string dtype here, special-casing whether or not you have numpy 1 or 2+ installed. That would prevent using paths longer than a certain amount but if I set the length to some high value all it would be fine for a couple of weeks and just waste a bit of memory...

Reply:
See also the test failures in https://github.com/zarr-developers/VirtualiZarr/pull/139, which I would rather fix once on this branch than twice (once in `main` and again on this branch).

Reply:
But... turns out in order to get this to pass the test suite I need to use a 128-bit dtype, in order to have enough characters to fit this extremely long temporary directory that pytest creates: `'/private/var/folders/zt/b84f41811p5_vygj1gjp6qx40000gn/T/pytest-of-tom/pytest-73/test_numpy_arrays_to_inlined_k0/air.nc'`.

Reply:
Does numpy really not allow variable-length strings?

What about using a pyarrow string array instead: https://arrow.apache.org/docs/python/generated/pyarrow.string.html

Reply:
> Does numpy really not allow variable-length strings?

A variable-length string dtype was one of the [headline features of numpy 2.0](https://numpy.org/devdocs/release/2.0.0-notes.html#highlights), released yesterday. I think the simplest thing to do here now is just to require numpy 2.0 as soon as possible (xarray already supports it, but I don't know about some packages like netcdf-python), and not worry about fixed-length strings at all.

> What about using a pyarrow string array instead:

I looked at this with @norlandrhagen, and it might yet be useful somewhere, but I think the numpy variable-length string dtype is exactly what we need for use inside the `Manifest`.

Reply:
https://github.com/zarr-developers/VirtualiZarr/pull/107/commits/8d2dcd5e452d192b1e931f037c624df8df9fabb2 just makes this PR depend on `numpy>=2.0.0`. I seem to now be able to build a local environment with all the dependencies I need and numpy 2 which is great.